ouroboros-consensus/src/Ouroboros/Consensus/MiniProtocol/ChainSync/Client.hs

```diff
@@ -937,6 +890,323 @@ chainSyncClient mkPipelineDecision0 tracer cfg
+          -- will abort this STM transaction.
+          check $ AF.length theirSuffix <= fromIntegral k
+          let toApply =
+                map headerRealPoint $
+                AF.toOldestFirst $
+                theirSuffix
+              ld      = LowDensity kis st toApply RequestStale
+          continueWithState ld kont
+  where
+    SecurityParam k = protocolSecurityParam (configConsensus cfg)
+    lCfg            = configLedger cfg
+
+-- | Acquire the requisite 'LedgerView' despite the peer having a low density
+-- chain.
+--
+-- TODO If we receive a RollForward that involves low density, and then the
+-- next reply from the server (NB the RequestNext was already sent due to
+-- pipelining) is a RollBackward, will we be able to process that, aborting our
+-- now-expensive low density RollForward handler? We don't think so. Is that
+-- OK? Seems suboptimal, but fine considering the narrow (disastrous)
+-- circumstances of a low density chain.
+lowDensityHandler :: forall blk m.
+     (MonadSTM m, LedgerSupportsProtocol blk)
+  => TopLevelConfig blk
+  -> ChainDbView m blk
+  -> (CandidateFragment (Header blk) -> m ())
+  -> SlotNo
+  -> Stateful (STM m) (LowDensity blk) (m (LedgerViewCheck blk))
+lowDensityHandler cfg cdbView setCandidate hdrSlot =
+    go1
+  where
+    ChainDbView {
+        getBlock
+      , getIsFetched
+      } = cdbView
+
+    lCfg = configLedger cfg
+
+    go1 :: Stateful (STM m) (LowDensity blk) (m (LedgerViewCheck blk))
+    go1 = Stateful worker
+
+    -- NOTE This binding has its own where clause.
+    worker ld = case toApply of
+
+        -- We have fetched and applied all blocks between the intersection
+        -- and the RollForward header.
+        [] ->
+            -- We can simply tick up to it, since there are no remaining
+            -- blocks in between.
+            return
+              $ return
+              $ LedgerViewAcquired kis
+              $ protocolLedgerView lCfg   -- project without forecasting
+              $ applyChainTick lCfg hdrSlot
+              $ st
+
+        -- Once the next block is fetched, advance the state and check if the
+        -- density is still too low.
+        rp:toApply' -> do
+            let p = realPointToPoint rp
+            -- TODO Will the blocks necessarily be in the restricted domain
+            -- of ChainDB.getIsFetched? We may need to widen it, or maybe
+            -- merely its current comment is too strong a statement.
+            isFetched <- getIsFetched <*> pure p
+
+            if
+
+              | isFetched -> return $ advanceAndRetryForecast rp toApply'
+
+              | RequestReady == reqStatus -> do
+                -- Block this STM transaction until the next block is fetched.
+                --
+                -- TODO what if the upstream peer is unable to serve the
+                -- block? Does BlockFetch silently ignore that? How could we
+                -- notice here?
```

Comment:
I guess the BlockFetch protocol will time out the connection with that peer. I assume that will also shut down the ChainSync protocol thread for the peer in question?